src/Dapr.Client/InvocationHandler.cs

```diff
@@ -0,0 +1,137 @@
+    {
+        private const string DefaultDaprHttpPort = "3500";
+
+        private static string GetDefaultDaprEndpoint()
+        {
+            var port = Environment.GetEnvironmentVariable("DAPR_HTTP_PORT") ?? DefaultDaprHttpPort;
+            return $"http://127.0.0.1:{port}";
+        }
+
+        private Uri parsedEndpoint;
+        private string? apiToken;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="InvocationHandler" />.
+        /// </summary>
+        public InvocationHandler()
+        {
+            this.parsedEndpoint = new Uri(GetDefaultDaprEndpoint(), UriKind.Absolute);
+            this.apiToken = Environment.GetEnvironmentVariable("DAPR_API_TOKEN");
+        }
+
+        /// <summary>
+        /// Gets or the sets the URI of the Dapr HTTP endpoint used for service invocation.
+        /// </summary>
+        /// <returns>The URI of the Dapr HTTP endpoint used for service invocation.</returns>
+        public string DaprEndpoint
+        {
+            get
+            {
+                return this.parsedEndpoint.OriginalString;
+            }
+            set
+            {
+                if (value == null)
+                {
+                    throw new ArgumentNullException(nameof(value));
+                }
+
+                // This will throw a reasonable exception if the URI is invalid.
+                this.parsedEndpoint = new Uri(value, UriKind.Absolute);
+            }
+        }
+
+        // Internal for testing
+        internal string? DaprApiToken
+        {
+            get => this.apiToken;
+            set => this.apiToken = value;
+        }
+
+        /// <inheritdoc />
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            var original = request.RequestUri;
+            if (!this.TryRewriteUri(request.RequestUri, out var rewritten))
+            {
+                throw new ArgumentException($"The request URI '{original}' is not a valid Dapr service invocation destination.", nameof(request));
+            }
+
+            try
+            {
+                if (this.apiToken is string)
+                {
+                    request.Headers.Add("dapr-api-token", this.apiToken);
+                }
+                request.RequestUri = rewritten;
+
+                return await base.SendAsync(request, cancellationToken);
+            }
+            finally
+            {
+                request.RequestUri = original;
+                request.Headers.Remove("dapr-api-token");
+            }
+        }
+
+        // Internal for testing
+        internal bool TryRewriteUri(Uri uri, [NotNullWhen(true)] out Uri? rewritten)
+        {
+            // For now the only invalid cases are when the request URI is missing or just silly.
+            // We may support additional cases for validation in the future (like an allow-list of App-Ids).
```

Comment:
Why should the allowlist be in the sdk and not dapr?

Reply:
Speculating about future feature ideas, but an allowlist at the SDK level would apply to the set of URLs we transform, not to access control.

It's a fairly common pattern with `HttpClient` where auth is involved to be able to specify what domains that auth applies to so you don't leak tokens to inappropriate domains.

A similar idea applies here... right now we assume this `HttpClient` is only used for Dapr service invocation. We might in the future add support for an allow list, so users could use the same HttpClient instance for non-Dapr cases.

Since this is effectively a TODO, there's no action to take here.